### PR TITLE
Allow gradients in fmap

### PR DESCRIPTION
**Pull request by DhairyaLGandhi**

### CODE REVIEW

**Comment by willtebbutt** [comment]
Any chance of some documentation alongside this PR explaining what the purpose of the two-argument version of `fmap` / `fmap1` is about.

Additionally, why do we need a separate `functor_tuple` function, as opposed to just implementing a method of `functor` for `Tuple`?

**Comment by DhairyaLGandhi** [comment]
Its to make Optimisers.jl work by returning a slightly trained version of the model. The tuple function is so that we return a tuple of layers in a chain. I should revisit to see if its necessary still but otherwise this should be merged so we can release Optimisers.jl

**Comment by willtebbutt** [comment]
I agree that this should probably go in, but a comment of some kind explaining what you've said above would be great.

File: src/functor.jl [hunk 1]
```diff
@@ -36,3 +54,8 @@ function fmap(f, x; cache = IdDict())
   haskey(cache, x) && return cache[x]
   cache[x] = isleaf(x) ? f(x) : fmap1(x -> fmap(f, x, cache = cache), x)
 end
+
```
Comment (DhairyaLGandhi) [inline src/functor.jl:57]:
```suggestion
# Allow gradients and other constructs that match the structure of the functor
# to allow for `map` style computations and return a modified version of the struct.
# This way we can use `fmap` to update the params with their gradients
```

Comment (DhairyaLGandhi) [reply src/functor.jl:57]: cc @willtebbutt

**Comment by DhairyaLGandhi** [comment]
bump!

File: src/functor.jl (hunk 1 above)
Comment (willtebbutt) [reply src/functor.jl:57]: LGTM!

**Review by willtebbutt — COMMENTED** [review]
The changes look get. Perhaps some tests and a patch bump?

**Review by ToucheSir — APPROVED** [review]

**Comment by ToucheSir** [comment]
Woops, meant to say LGTM once the tests land.

**Comment by DhairyaLGandhi** [comment]
Allows writing something like

```julia
Functors.fmap(x, y) do x, y
  isnothing(x) && return y
  isnothing(y) && return x
  f(x,y)
end
```

where x and y are basically NamedTuples out of Zygote, or a mix of the actual object tree (think Chain, model struct etc) and the corresponding gradients.

File: src/functor.jl [hunk 2]
```diff
@@ -44,10 +44,29 @@ Equivalent to `functor(x)[1]`.
 """
 children(x) = functor(x)[1]
 
+function functor_tuple(f, x::Tuple, dx::Tuple)
```
Comment (darsnack) [inline src/functor.jl:47]: I don't think `x` is a tuple
```suggestion
function functor_tuple(f, x, dx::Tuple)
```

File: src/functor.jl [hunk 3]
```diff
@@ -44,10 +44,29 @@ Equivalent to `functor(x)[1]`.
...
+end
+functor_tuple(f, x, dx) = f(x, dx)
+functor_tuple(f, x, ::Nothing) = x
+
+# @functor Chain
+# Chain -> func = (layers = (Dense,Dense),), gs -> (layers...)
+function _default_walk(f, x, dx)
+  func, re = functor(x)
+  map(func, dx) do x, x̄
+    # functor_tuple(f, x, x̄)
```
Comment (darsnack) [inline src/functor.jl:60]: This should work with the other change?

**Review by darsnack — COMMENTED** [review]
Looks good!

**Comment by ToucheSir** [comment]
@DhairyaLGandhi this already has my approval, so feel free to merge any time :)

**Pull request merged**